raster/src/main/scala/geotrellis/raster/io/geotiff/GeoTiffMultibandTile.scala

```diff
@@ -321,8 +324,145 @@ abstract class GeoTiffMultibandTile(
   /**
     * Converts the GeoTiffMultibandTile to an
     * [[ArrayMultibandTile]] */
-  def toArrayTile(): ArrayMultibandTile =
-    ArrayMultibandTile(bands.map(_.toArrayTile): _*)
+  def toArrayTile(): ArrayMultibandTile = {
+    val tiles = Array.fill[MutableArrayTile](bandCount)(ArrayTile.empty(cellType, cols, rows))
+
+    val tileSetInt = (
+      segment: GeoTiffSegment,
+      segmentTransform: SegmentTransform,
+      bandIndex: Int
+    ) => {
+      cfor(0)(_ < segment.size, _ + 1) { i =>
+        val col = segmentTransform.indexToCol(i)
+        val row = segmentTransform.indexToRow(i)
+        if (col < cols && row < rows) {
+          val v = segment.getInt(i)
+          tiles(bandIndex).set(col, row, v)
+        }
+      }
+    }
+
+    val tileSetDouble = (
+      segment: GeoTiffSegment,
+      segmentTransform: SegmentTransform,
+      bandIndex: Int
+    ) => {
+      cfor(0)(_ < segment.size, _ + 1) { i =>
+        val col = segmentTransform.indexToCol(i)
+        val row = segmentTransform.indexToRow(i)
+        if (col < cols && row < rows) {
+          val v = segment.getDouble(i)
+          tiles(bandIndex).setDouble(col, row, v)
+        }
+      }
+    }
+
+    if (hasPixelInterleave) {
+      bandType match {
+        case BitBandType =>
+          getSegments(0 until segmentCount).foreach { case (segmentIndex, segment) =>
+            val (segCols, segRows) =
+              if (segmentLayout.isTiled) (segmentLayout.tileLayout.tileCols, segmentLayout.tileLayout.tileRows)
+              else segmentLayout.getSegmentDimensions(segmentIndex)
+
+            val segmentTransform = segmentLayout.getSegmentTransform(segmentIndex, 1)
+            val bytes = GeoTiffSegment.deinterleaveBitSegment(segment, segCols, segRows, bandCount, 0 until bandCount)
+
+            cfor(0)(_ < bandCount, _ + 1) { bandIndex =>
+              val segmentBytes = bytes(bandIndex)
+              tileSetInt(new BitGeoTiffSegment(segmentBytes, segCols, segRows), segmentTransform, bandIndex)
+            }
+          }
+        case _ =>
+          val segmentProducer =
```

Comment:
I'm wondering can we abstract over it? I mean to avoid somehow strict types

Reply:
Btw, if there is no way to do it, probably we can move that segmentProducer function into `GeoTiffSegment` object.